spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -101,15 +118,50 @@ private void validateAndSetFilterParts() {
 
 	@Override
 	public Object execute(Object[] parameters) {
+		Class<?> returnedObjectType = getQueryMethod().getReturnedObjectType();
+		Object result;
+		if (isSliceQuery() || isPageQuery()) {
+			List<?> resultEntries = (List) execute(parameters, returnedObjectType, List.class, false);
+			Long totalCount = (Long) execute(parameters, Long.class, null, true);
+
+			ParameterAccessor paramAccessor =
+					new ParametersParameterAccessor(getQueryMethod().getParameters(), parameters);
+
+			Pageable pageable = paramAccessor.getPageable();
+			if (getQueryMethod().isSliceQuery()) {
+				boolean hasNext = !pageable.equals(Pageable.unpaged()) && pageable.next().getOffset() < totalCount;
+				result = new SliceImpl<>(resultEntries, pageable, hasNext);
+			}
+			else {
+				result = new PageImpl<>(resultEntries, pageable, totalCount);
```

Comment:
Since Page is a child of Slice, could you just use PageImpl for both slice and page?

Reply:
Excellent point!